### PR TITLE
storage_service: Enable RBNO by default for replace

### PR DESCRIPTION
**Pull request by asias**
We decided to enable repair based node operations by default for replace
node operations.

To do that, a new option --repair-based-node-ops-whitelist is added. It
lists the node operations that are allowed to enable repair based node
operations.

The operations can be bootstrap, replace, removenode, decommission and rebuild.

By default, --repair-based-node-ops-whitelist is set to contain "replace".

Note, the existing option --enable-repair-based-node-ops is still in
play. It is the global switch to enable or disable the feature.

Examples:

- To enable bootstrap and replace node ops:

```
scylla --enable-repair-based-node-ops true --repair-based-node-ops-whitelist replace,bootstrap
```

- To disable any repair based node ops:

```
scylla --enable-repair-based-node-ops false
```

### CODE REVIEW

**Comment by asias** [comment]
@slivne @avikivity @roydahan  FYI

**Comment by dorlaor** [comment]
Congratulations Asias! Hope to see the rest become the default later.
One enhancement that I like to ask is to dynamically switch between
the modes. So if I'm using bootstrap RNBO when I booted the cluster,
I could still change the mode to the old style streaming w/o a reboot

**Comment by asias** [comment]
> Congratulations Asias! Hope to see the rest become the default later. One enhancement that I like to ask is to dynamically switch between the modes. So if I'm using bootstrap RNBO when I booted the cluster, I could still change the mode to the old style streaming w/o a reboot

Thanks. Me too. It works this way already. The option is liveness::LiveUpdate, so no reboot is needed after one changes the option. And I intentionally parse the option every time is_repair_based_node_ops_enabled is called so it will use the latest config without a reboot.

File: db/config.cc [hunk 1]
```diff
@@ -737,7 +737,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , replace_address(this, "replace_address", value_status::Used, "", "The listen_address or broadcast_address of the dead node to replace. Same as -Dcassandra.replace_address.")
     , replace_address_first_boot(this, "replace_address_first_boot", value_status::Used, "", "Like replace_address option, but if the node has been bootstrapped successfully it will be ignored. Same as -Dcassandra.replace_address_first_boot.")
     , override_decommission(this, "override_decommission", value_status::Used, false, "Set true to force a decommissioned node to join the cluster")
-    , enable_repair_based_node_ops(this, "enable_repair_based_node_ops", liveness::LiveUpdate, value_status::Used, false, "Set true to use enable repair based node operations instead of streaming based")
+    , repair_based_node_ops_whitelist(this, "repair_based_node_ops_whitelist", liveness::LiveUpdate, value_status::Used, "replace", "A comma separated list of node operations which are allowed to enable repair based node operations. The operations can be bootstrap, replace, removenode, decommission and rebuild")
```
[the remark this answers is not in the document]
Comment (asias) [reply db/config.cc:740]: OK. I will change the word. I thought about changing the meaning of enable_repair_based_node_ops initially but I dropped it.
If we do this, we need to chagne the default to  --enable-repair-based-node-ops "replace", however we have the old config --enable-repair-based-node-ops true in our tests, we need to change all those tests. When we allow another operation, we need to do it again. 

With a new config, we do not need to change anything. Over time, the list will be all the operations, we will forget about the list option. 

I will change the name to allowed_repair_based_node_ops.

[the remark this answers is not in the document]
Comment (avikivity) [reply db/config.cc:740]: ok.

[the remark this answers is not in the document]
Comment (asias) [reply db/config.cc]: Fixed in v2.

**Comment by asias** [comment]
Changes in v2:
- Rename the option to  --allowed-repair-based-node-ops

**Pull request closed without merging**